File: airbyte-commons-worker/src/test/java/io/airbyte/workers/StateWithIdTest.java
/*
 * Copyright (c) 2020-2025 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.workers;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotEquals;

import com.fasterxml.jackson.databind.JsonNode;
import io.airbyte.commons.json.Jsons;
import io.airbyte.protocol.models.AirbyteGlobalState;
import io.airbyte.protocol.models.AirbyteMessage;
import io.airbyte.protocol.models.AirbyteStateMessage;
import io.airbyte.protocol.models.AirbyteStateStats;
import io.airbyte.protocol.models.AirbyteStreamState;
import io.airbyte.protocol.models.StreamDescriptor;
import io.airbyte.workers.models.StateWithId;
import java.util.Collections;
import java.util.Optional;
import java.util.Random;
import java.util.UUID;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;
import org.testcontainers.shaded.com.google.common.collect.ImmutableMap;

class StateWithIdTest {

  @ParameterizedTest
  @ValueSource(strings = {
    "{\"{\\\"schema\\\":null,\\\"payload\\\":[\\\"db_jagkjrgxhw\\\",{\\\"server\\\":\\\"db_jagkjrgxhw\\\"}]}\":"
        + "\"{\\\"last_snapshot_record\\\":true,\\\"lsn\\\":23896935,\\\"txId\\\":505,\\\"ts_usec\\\":1659422332985000,\\\"snapshot\\\":true}\"}",
    "{\"[\\\"db_jagkjrgxhw\\\",{\\\"server\\\":\\\"db_jagkjrgxhw\\\"}]}\":\"{\\\"last_snapshot_record\\\":true,"
        + "\\\"lsn\\\":23896935,\\\"txId\\\":505,\\\"ts_usec\\\":1659422332985000,\\\"snapshot\\\":true}\"}",
    "{\"[\\\"db_jagkjrgxhw\\\",{\\\"server\\\":\\\"db_jagkjrgxhw\\\"}]\":\"{\\\"transaction_id\\\":null,\\\"lsn\\\":"
        + "23896935,\\\"txId\\\":505,\\\"ts_usec\\\":1677520006097984}\"}"
  })
  void globalStateTest(final String cdcState) {
    final Random random = new Random();
    final double recordCount = random.nextDouble();
    final String cursorName = UUID.randomUUID().toString();

    final AirbyteStateMessage originalState = getAirbyteGlobalStateMessage(cdcState, recordCount, cursorName);
    final AirbyteMessage originalMessage = new AirbyteMessage().withType(AirbyteMessage.Type.STATE)
        .withState(originalState);

    final AirbyteStateMessage copyOfOriginalState = getAirbyteGlobalStateMessage(cdcState, recordCount, cursorName);
    final AirbyteMessage copyOfOriginal = new AirbyteMessage().withType(AirbyteMessage.Type.STATE)
        .withState(copyOfOriginalState);

    assertEquals(originalMessage, copyOfOriginal);
    final AirbyteMessage stateMessageWithIdAdded = StateWithId.attachIdToStateMessageFromSource(copyOfOriginal);
    assertNotEquals(originalMessage, stateMessageWithIdAdded);
    assertEquals(originalMessage.getState().getGlobal(), stateMessageWithIdAdded.getState().getGlobal());

    final String serializedMessage = Jsons.serialize(stateMessageWithIdAdded);
    Optional<AirbyteMessage> deserializedMessage = Jsons.tryDeserializeExact(serializedMessage, AirbyteMessage.class);
    assertEquals(stateMessageWithIdAdded, deserializedMessage.orElseThrow());
    assertEquals(originalMessage.getState().getGlobal(), deserializedMessage.orElseThrow().getState().getGlobal());
    assertEquals(StateWithId.getIdFromStateMessage(stateMessageWithIdAdded).orElseThrow(),
        StateWithId.getIdFromStateMessage(deserializedMessage.orElseThrow()).orElseThrow());
  }

  @Test
  void streamStateTest() {
    final Random random = new Random();
    final double recordCount = random.nextDouble();
    final String cursorName = UUID.randomUUID().toString();

    final AirbyteStateMessage originalState = getAirbyteStreamStateMessage(recordCount, cursorName);
    final AirbyteMessage originalMessage = new AirbyteMessage().withType(AirbyteMessage.Type.STATE)
        .withState(originalState);

    final AirbyteStateMessage copyOfOriginalState = getAirbyteStreamStateMessage(recordCount, cursorName);
    final AirbyteMessage copyOfOriginal = new AirbyteMessage().withType(AirbyteMessage.Type.STATE)
        .withState(copyOfOriginalState);

    assertEquals(originalMessage, copyOfOriginal);
    final AirbyteMessage stateMessageWithIdAdded = StateWithId.attachIdToStateMessageFromSource(copyOfOriginal);
    assertNotEquals(originalMessage, stateMessageWithIdAdded);
    assertEquals(originalMessage.getState().getGlobal(), stateMessageWithIdAdded.getState().getGlobal());

    final String serializedMessage = Jsons.serialize(stateMessageWithIdAdded);
    Optional<AirbyteMessage> deserializedMessage = Jsons.tryDeserializeExact(serializedMessage, AirbyteMessage.class);
    assertEquals(stateMessageWithIdAdded, deserializedMessage.orElseThrow());
    assertEquals(originalMessage.getState().getStream(), deserializedMessage.orElseThrow().getState().getStream());
    assertEquals(StateWithId.getIdFromStateMessage(stateMessageWithIdAdded).orElseThrow(),
        StateWithId.getIdFromStateMessage(deserializedMessage.orElseThrow()).orElseThrow());
  }

  private static AirbyteStateMessage getAirbyteStreamStateMessage(final double recordCount, final String cursorName) {
    final AirbyteStreamState streamState = new AirbyteStreamState()
        .withStreamState(Jsons.jsonNode(ImmutableMap.of(cursorName, 1)))
        .withStreamDescriptor(new StreamDescriptor().withName(cursorName).withNamespace(cursorName));
    final AirbyteStateStats airbyteStateStats = new AirbyteStateStats().withRecordCount(recordCount);
    return new AirbyteStateMessage()
        .withType(AirbyteStateMessage.AirbyteStateType.STREAM)
        .withStream(streamState)
        .withSourceStats(airbyteStateStats);
  }

  private static AirbyteStateMessage getAirbyteGlobalStateMessage(final String cdcState, final double recordCount, final String cursorName) {
    final JsonNode cdcStateAsJson = Jsons.deserialize(cdcState);
    final AirbyteGlobalState globalState = new AirbyteGlobalState().withSharedState(cdcStateAsJson).withStreamStates(Collections.singletonList(
        new AirbyteStreamState()
            .withStreamState(Jsons.jsonNode(ImmutableMap.of(cursorName, 1)))
            .withStreamDescriptor(new StreamDescriptor().withName(cursorName).withNamespace(cursorName))));
    final AirbyteStateStats airbyteStateStats = new AirbyteStateStats().withRecordCount(recordCount);
    return new AirbyteStateMessage()
        .withType(AirbyteStateMessage.AirbyteStateType.GLOBAL)
        .withGlobal(globalState)
        .withSourceStats(airbyteStateStats);
  }

}

File: airbyte-container-orchestrator/src/test/kotlin/io/airbyte/container/orchestrator/worker/model/StateWithIdTest.kt
/*
 * Copyright (c) 2020-2025 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.container.orchestrator.worker.model

import io.airbyte.commons.json.Jsons
import io.airbyte.protocol.models.v0.AirbyteGlobalState
import io.airbyte.protocol.models.v0.AirbyteMessage
import io.airbyte.protocol.models.v0.AirbyteStateMessage
import io.airbyte.protocol.models.v0.AirbyteStateStats
import io.airbyte.protocol.models.v0.AirbyteStreamState
import io.airbyte.protocol.models.v0.StreamDescriptor
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNotEquals
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.ValueSource
import java.util.Random
import java.util.UUID

internal class StateWithIdTest {
  @ParameterizedTest
  @ValueSource(
    strings = [
      (
        "{\"{\\\"schema\\\":null,\\\"payload\\\":[\\\"db_jagkjrgxhw\\\",{\\\"server\\\":\\\"db_jagkjrgxhw\\\"}]}\":" +
          "\"{\\\"last_snapshot_record\\\":true,\\\"lsn\\\":23896935,\\\"txId\\\":505,\\\"ts_usec\\\":1659422332985000,\\\"snapshot\\\":true}\"}"
      ), (
        "{\"[\\\"db_jagkjrgxhw\\\",{\\\"server\\\":\\\"db_jagkjrgxhw\\\"}]}\":\"{\\\"last_snapshot_record\\\":true," +
          "\\\"lsn\\\":23896935,\\\"txId\\\":505,\\\"ts_usec\\\":1659422332985000,\\\"snapshot\\\":true}\"}"
      ), (
        "{\"[\\\"db_jagkjrgxhw\\\",{\\\"server\\\":\\\"db_jagkjrgxhw\\\"}]\":\"{\\\"transaction_id\\\":null,\\\"lsn\\\":" +
          "23896935,\\\"txId\\\":505,\\\"ts_usec\\\":1677520006097984}\"}"
      ),
    ],
  )
  fun globalStateTest(cdcState: String?) {
    val random = Random()
    val recordCount = random.nextDouble()
    val cursorName = UUID.randomUUID().toString()

    val originalState = getAirbyteGlobalStateMessage(cdcState, recordCount, cursorName)
    val originalMessage =
      AirbyteMessage()
        .withType(AirbyteMessage.Type.STATE)
        .withState(originalState)

    val copyOfOriginalState = getAirbyteGlobalStateMessage(cdcState, recordCount, cursorName)
    val copyOfOriginal =
      AirbyteMessage()
        .withType(AirbyteMessage.Type.STATE)
        .withState(copyOfOriginalState)

    assertEquals(originalMessage, copyOfOriginal)
    val stateMessageWithIdAdded: AirbyteMessage = attachIdToStateMessageFromSource(copyOfOriginal)
    assertNotEquals(originalMessage, stateMessageWithIdAdded)
    assertEquals(originalMessage.state.global, stateMessageWithIdAdded.state.global)

    val serializedMessage = Jsons.serialize(stateMessageWithIdAdded)
    val deserializedMessage = Jsons.tryDeserializeExact(serializedMessage, AirbyteMessage::class.java)
    assertEquals(stateMessageWithIdAdded, deserializedMessage.orElseThrow())
    assertEquals(originalMessage.state.global, deserializedMessage.orElseThrow()!!.state.global)
    assertEquals(
      getIdFromStateMessage(stateMessageWithIdAdded),
      getIdFromStateMessage(deserializedMessage.orElseThrow()),
    )
  }

  @Test
  fun streamStateTest() {
    val random = Random()
    val recordCount = random.nextDouble()
    val cursorName = UUID.randomUUID().toString()

    val originalState = getAirbyteStreamStateMessage(recordCount, cursorName)
    val originalMessage =
      AirbyteMessage()
        .withType(AirbyteMessage.Type.STATE)
        .withState(originalState)

    val copyOfOriginalState = getAirbyteStreamStateMessage(recordCount, cursorName)
    val copyOfOriginal =
      AirbyteMessage()
        .withType(AirbyteMessage.Type.STATE)
        .withState(copyOfOriginalState)

    assertEquals(originalMessage, copyOfOriginal)
    val stateMessageWithIdAdded: AirbyteMessage = attachIdToStateMessageFromSource(copyOfOriginal)
    assertNotEquals(originalMessage, stateMessageWithIdAdded)
    assertEquals(originalMessage.state.global, stateMessageWithIdAdded.state.global)

    val serializedMessage = Jsons.serialize(stateMessageWithIdAdded)
    val deserializedMessage = Jsons.tryDeserializeExact(serializedMessage, AirbyteMessage::class.java)
    assertEquals(stateMessageWithIdAdded, deserializedMessage.orElseThrow())
    assertEquals(originalMessage.state.stream, deserializedMessage.orElseThrow()!!.state.stream)
    assertEquals(
      getIdFromStateMessage(stateMessageWithIdAdded),
      getIdFromStateMessage(deserializedMessage.orElseThrow()),
    )
  }

  @Test
  fun testAttachStateId() {
    val stateMessage = AirbyteStateMessage()
    val updatedStateMessage = attachIdToStateMessageFromSource(stateMessage)
    assertTrue(updatedStateMessage.additionalProperties.contains(ID))
    val updatedStateMessage2 = attachIdToStateMessageFromSource(updatedStateMessage)
    assertEquals(updatedStateMessage.additionalProperties[ID], updatedStateMessage2.additionalProperties[ID])
  }

  @Test
  fun testMissingStateId() {
    val message = AirbyteStateMessage()
    assertThrows(IllegalStateException::class.java) {
      getIdFromStateMessage(message)
    }
  }

  private fun getAirbyteStreamStateMessage(
    recordCount: Double,
    cursorName: String,
  ): AirbyteStateMessage? {
    val streamState =
      AirbyteStreamState()
        .withStreamState(Jsons.jsonNode(mapOf(cursorName to 1)))
        .withStreamDescriptor(StreamDescriptor().withName(cursorName).withNamespace(cursorName))
    val airbyteStateStats = AirbyteStateStats().withRecordCount(recordCount)
    return AirbyteStateMessage()
      .withType(AirbyteStateMessage.AirbyteStateType.STREAM)
      .withStream(streamState)
      .withSourceStats(airbyteStateStats)
  }

  private fun getAirbyteGlobalStateMessage(
    cdcState: String?,
    recordCount: Double,
    cursorName: String,
  ): AirbyteStateMessage? {
    val cdcStateAsJson = Jsons.deserialize(cdcState)
    val globalState =
      AirbyteGlobalState().withSharedState(cdcStateAsJson).withStreamStates(
        listOf<AirbyteStreamState?>(
          AirbyteStreamState()
            .withStreamState(Jsons.jsonNode(mapOf(cursorName to 1)))
            .withStreamDescriptor(StreamDescriptor().withName(cursorName).withNamespace(cursorName)),
        ),
      )
    val airbyteStateStats = AirbyteStateStats().withRecordCount(recordCount)
    return AirbyteStateMessage()
      .withType(AirbyteStateMessage.AirbyteStateType.GLOBAL)
      .withGlobal(globalState)
      .withSourceStats(airbyteStateStats)
  }
}
